### ISSUE TITLE
Allow some circular dependencies?

### ISSUE THREAD

**Issue by frank06** [rrousselGit/river_pod#35]
**Is your feature request related to a problem? Please describe.**

Given data classes of `Post` and `Comment`, Riverpod works great wiring up their repositories' "boot" dependencies (a tree-like structure):

```
hiveProvider <- postsHiveBoxProvider <- postsRepositoryProvider

hiveProvider <- commentsHiveBoxProvider <- commentsRepositoryProvider
```

`Post`s, `Comment`s and other models, however, depend on each other. These relationships are better represented with a directed graph data structure, which can be navigated like:

```
final post = await ref.read(postsRepositoryProvider).getPost();
post.comments.first.post.value.comments.first...
```

Since `postsRepository` needs to hold a reference to `commentsRepository`, once `ref.read(commentsRepositoryProvider)` is called I get a `CircularDependencyError`.

**Describe the solution you'd like**

Allow disabling the circular reference assertion for legitimate cases like Providers in graph-like structures.

```
ref.read(commentsRepositoryProvider, circular: true)
```

Perhaps there are alternatives (using cached providers)? I not too familiar with Riverpod

**Comment by rrousselGit** [comment]
Circular dependencies are not allowed for performance and simplicity reasons – on the top of the maintainability bonuses. 
As such, they will never be allowed.

But there should be an alternative solution for your case. For example, you could have:

                              hide
                              /    \
              comments     posts
                               \   /
                          messages

It's hard to say without knowing exactly what you want. But in any case, circular dependencies are not the solution
Feel free to share about your use-case so that I can help you with it.

**Comment by frank06** [comment]
Thanks Remi! Sorry that your comment slipped through the cracks. Appreciate the help!

I didn't quite get your example. I went back to the project and split up `Repository` which had one responsibility too many. Cyclic deps gone.

Fwiw I found `ref.read` a bit confusing when passing it around/working with many dependencies (esp since `read` is used throughout rp in different contexts). I started being more specific with refs and used an alias, i.e. `repositoryRef.readDependency(x)`
